ff/src/fields/models/quadratic_extension.rs

```diff
@@ -216,34 +238,53 @@ impl<P: QuadExtParameters> Field for QuadExtField<P> {
     }
 
     fn square_in_place(&mut self) -> &mut Self {
-        // v0 = c0 - c1
-        let mut v0 = self.c0 - &self.c1;
-        // v3 = c0 - beta * c1
-        let v3 = self.c0 - &P::mul_base_field_by_nonresidue(&self.c1);
-        // v2 = c0 * c1
-        let v2 = self.c0 * &self.c1;
+        if P::NONRESIDUE == -P::BaseField::one() {
+            // Optimization when the non-residue is -1, where we can remove 2 intermediate additions
+            // and use one fewer intermediate
+            let c0_copy = self.c0;
+            // v0 = c0 - c1
+            let v0 = self.c0 - &self.c1;
+            // result.c1 = 2 c1
+            self.c1.double_in_place();
+            // result.c0 = (c0 - c1) + 2c1 = c0 + c1
+            self.c0 = v0 + &self.c1;
+            // result.c0 *= (c0 - c1)
+            // result.c0 = (c0 + c1) * (c0 - c1) = c0^2 - c1^2
+            self.c0 *= v0;
+            // result.c1 *= c0
+            // result.c1 = 2 * c1 * c0
+            self.c1 *= c0_copy;
 
-        // v0 = (v0 * v3) + v2
-        v0 *= &v3;
-        v0 += &v2;
+            self
```

Comment:
I think it's worthwhile here to add the complex multiplication formula, and then show how it's optimized in the `P::NONRESIDUE == -1` case.

Reply:
Added, this is a great suggestion! Also it pointed out another field op we can save in squaring for bls12_377